### ISSUE TITLE
DF description/missing color is unreadable on dark themes

### ISSUE THREAD

**Issue by fbruetting** [JuliaData/DataFrames.jl#3400]
In the picture below you see a DataFrame in combination with following themes:
  * Solarized Dark
  * Solarized Dark (with color “alternate black” manually changed to grey)
  * Solarized Light

![Bildschirmfoto vom 2023-11-08 17-54-58](https://github.com/JuliaData/DataFrames.jl/assets/10520767/1b370cde-9ba1-4f9a-9ac5-3e649b683762)

Solarized Dark is one of the most popular dark themes, because of its very good readability as its contrast being neither too low nor too high. Without modifications, `missing`s and column types are not readable at all.

I found out that this is due to DataFrames using the “alternate black” color – which on dark themes usually is equal to the background color, or at least very near. So it’s not a problem at all on light themes, but all dark themes have problems with this (not just Solarized). See the following examples with the themes:
  * Adwaita Dark
  * Builder Dark (GNOME Builder)
  * Classic Dark
  * Cobalt
  * Kate Dark
  * Peninsula Dark
  * Solarized Dark
  * Oblivion
  * Arctic Dark

I haven’t encountered these issues anywhere else, neither in my IDE (GNOME Builder), nor in general Julia, so I guess the problem’s root is this package’s color selection. I guess it’s necessary to chose another color here, as “black” just works well one bright themes.

![Bildschirmfoto vom 2023-11-08 17-59-02](https://github.com/JuliaData/DataFrames.jl/assets/10520767/ebcf0c9c-e78d-4c07-aaba-75b51bd103f1)
![Bildschirmfoto vom 2023-11-08 17-59-05](https://github.com/JuliaData/DataFrames.jl/assets/10520767/78de072a-b2e5-4b0d-a08e-56997f6b3470)
![Bildschirmfoto vom 2023-11-08 17-59-07](https://github.com/JuliaData/DataFrames.jl/assets/10520767/c79bb7e1-41d3-4482-a794-0b87a0c597ef)
![Bildschirmfoto vom 2023-11-08 17-59-09](https://github.com/JuliaData/DataFrames.jl/assets/10520767/6715be4e-780c-4c74-b7a7-0efdfc90a79a)
![Bildschirmfoto vom 2023-11-08 17-59-11](https://github.com/JuliaData/DataFrames.jl/assets/10520767/16e0dee4-3c07-40d8-ad67-a8ada448e91d)
![Bildschirmfoto vom 2023-11-08 17-59-12](https://github.com/JuliaData/DataFrames.jl/assets/10520767/87dc85a6-613e-4c8e-8ebb-cbe27d64f39e)
![Bildschirmfoto vom 2023-11-08 17-59-14](https://github.com/JuliaData/DataFrames.jl/assets/10520767/7e35e1c2-5a58-4761-b0d7-19ac5eefd1ac)
![Bildschirmfoto vom 2023-11-08 17-59-16](https://github.com/JuliaData/DataFrames.jl/assets/10520767/8f691a47-4f52-4f8e-964b-819c3939f83e)
![Bildschirmfoto vom 2023-11-08 17-59-18](https://github.com/JuliaData/DataFrames.jl/assets/10520767/35e0399c-ed68-40e1-ae99-9028ec64f19e)

**Comment by bkamins** [comment]
Thank you for the comment. Do you have a recommendation then?

CC @ronisbr

**Comment by fbruetting** [comment]
Thank for the quick reply!

I think one of the colours should be chosen, as neither `background`/`foreground` nor `black`/`dark grey`/`light gray`/`white` seem to work well for both light an dark themes.

This leaves `red`, `green`, `orange`, `blue`, `magenta` and `cyan` as options. The alternate colours originally were intended as brighter tones (`light red`, `light green`, `yellow`, `light blue`, `light purple` and `teal`), but some themes – for whatever reason – abuse them for completely different colours. Solarized Dark for example uses these places in the colour array to store some dark tones. I don’t know why this is the case, but it seems these cannot be used reliably in every case. So if you chose for example `yellow`, then in some themes that text gets flashy, while in Solarized Dark it gets very dark.

[This webpage](https://opensource.com/article/19/9/linux-terminal-colors) might be helpful.

Here are some theme examples:

![Bildschirmfoto vom 2023-11-08 20-00-04](https://github.com/JuliaData/DataFrames.jl/assets/10520767/56860136-27c5-4cb9-8b72-72d90a1d53dc)
![Bildschirmfoto vom 2023-11-08 20-00-19](https://github.com/JuliaData/DataFrames.jl/assets/10520767/f5ce738c-adbb-4e7c-af8f-bbe0a04b4b31)
![Bildschirmfoto vom 2023-11-08 20-00-27](https://github.com/JuliaData/DataFrames.jl/assets/10520767/3576599b-4808-4472-94c0-3aff1fec3e44)
![Bildschirmfoto vom 2023-11-08 20-00-31](https://github.com/JuliaData/DataFrames.jl/assets/10520767/4cc219a5-3379-4b0c-98bb-da51afd97755)
![Bildschirmfoto vom 2023-11-08 20-00-36](https://github.com/JuliaData/DataFrames.jl/assets/10520767/33bea413-a010-4ded-aab1-e3e2972413ec)
![Bildschirmfoto vom 2023-11-08 20-00-43](https://github.com/JuliaData/DataFrames.jl/assets/10520767/1e4cefce-8c05-42f6-b882-e8caf0d56fe2)
![Bildschirmfoto vom 2023-11-08 20-00-48](https://github.com/JuliaData/DataFrames.jl/assets/10520767/f2b60330-0c9a-48f4-9deb-31ea640b17b1)

**Comment by ronisbr** [comment]
Hi!

Fixing this issue would be extremely simple since we just need to change the color here:

https://github.com/JuliaData/DataFrames.jl/blob/8c32d537b1d5a2ae9b1cdd72928def18aeae0bf9/src/abstractdataframe/prettytables.jl#L28C48-L28C79

Personally, I have no idea what would be the best option.

**Comment by fbruetting** [comment]
The best option, I guess, would be to select `light gray` for dark themes and `dark gray` for light themes. But that seems complicated. 😶️

**Comment by bkamins** [comment]
In summary: we are OK to change this given we know what color to pick. The restriction is that:
1. It has to be one color (as we cannot detect the theme the user uses).
2. The intention is to dim `missing` a bit (it should not cry-out, rather it should be visually a bit less standing out, as this is the point of having a different color for it)

**Comment by ronisbr** [comment]
Hum, there is no universal way to check if the theme is dark or light AFAIK :(

**Comment by ronisbr** [comment]
I agree 100% with @bkamins 's proposal here.

**Comment by fbruetting** [comment]
> In summary: we are OK to change this given we know what color to pick. The restriction is that:
> 
>     1. It has to be one color (as we cannot detect the theme the user uses).
> 
>     2. The intention is to dim `missing` a bit (it should not cry-out, rather it should be visually a bit less standing out, as this is the point of having a different color for it)

I searched for a bit, but it doesn’t seem possible to tick both boxes. Maybe then hardcoding a medium gray, which is suitable for both theme variants?

**Comment by ronisbr** [comment]
> I searched for a bit, but it doesn’t seem possible to tick both boxes. Maybe then hardcoding a medium gray, which is suitable for both theme variants?

Do you mean, setting a RGB color?

**Comment by bkamins** [comment]
I think it would make sense to hardcode RGB medium gray (I guess it is possible - right?)

**Comment by fbruetting** [comment]
In the shell it sure is, I just don’t know how Julia interfaces with that. Shells use strange color codes like `\033[31m`, as you can see in my link above.

**Comment by ronisbr** [comment]
The problem is for terminals that do not support 24-bit colors. I need to test to check what will happen.

**Comment by ronisbr** [comment]
> In the shell it sure is, I just don’t know how Julia interfaces with that. Shells use strange color codes like `\033[31m`, as you can see in my link above.

This code says the terminal must change the foreground color to red. In PrettyTables, we use Crayons.jl that is a conversor between a color and those escape sequences.

Maybe we can assume everyone will be using at least a terminal that supports 256 colors. In this case, we have access to a wider range of colors. IMHO, assume 24-bit color is too much restrictive.

In the case of 256 colors, there are those options:

<img width="829" alt="IMG_4384" src="https://github.com/JuliaData/DataFrames.jl/assets/1068295/a737ebdf-336c-4aab-886f-697a4a83bf50">

(from https://www.lihaoyi.com/post/BuildyourownCommandLinewithANSIescapecodes.html)

Thus, we can select a color between 232 and 255.

**Comment by fbruetting** [comment]
You can also substitute `:dark_gray` by `(127,127,127)` in your referenced code.

**Comment by ronisbr** [comment]
But AFAIK it is converted to a 24-bit color code, which should not work in terminals that do not support it.

**Comment by fbruetting** [comment]
While we’re at this: Is it possible to specify this color as an environment variable or by a function call? I guess that would be nice for circumventing corner cases due to hardcoded values, for example in the rare case anyone has a grey background.

**Comment by ronisbr** [comment]
Yes, it should be fine to check if a ENV variable is defined and use this value instead. Maybe it will add an overhead, but should not be noticeable. What do you think @bkamins ? Any suggestions about the variable name?

**Comment by bkamins** [comment]
In the past we have had a policy not to introduce behavior depending on global state.

@nalimilan - in this case do you think it would be acceptable?

**Comment by giordano** [comment]
Isn't this the same problem as https://github.com/JuliaLang/julia/issues/38730 caused by the fact the solarized theme uses same colours for background and foreground?

**Comment by fbruetting** [comment]
> Isn't this the same problem as [JuliaLang/julia#38730](https://github.com/JuliaLang/julia/issues/38730) caused by the fact the solarized theme uses same colours for background and foreground?

Yup, seems so. Thanks for the tip, now I see that my stack traces contain more information than what is visible. 😅️

**Comment by nalimilan** [comment]
> In the past we have had a policy not to introduce behavior depending on global state.
> 
> @nalimilan - in this case do you think it would be acceptable?

I would find it kind of weird to have a mechanism to set the color used when printing a particular object (`missing`). Isn't there a more general theming system that we could use?

Anyway, regarding this particular issue, requiring users to set an environment variable seems suboptimal. Most of them won't even know they can do that. Is there a way to check whether `light_grey` gives the same color as the background, and switch to `dark_grey` in that case?

**Comment by ronisbr** [comment]
Unfortunately no. There is no universal way to check what is the color the terminal will use given a ANSI escape sequence.

**Comment by fbruetting** [comment]
> > In the past we have had a policy not to introduce behavior depending on global state.
> > @nalimilan - in this case do you think it would be acceptable?
> 
> I would find it kind of weird to have a mechanism to set the color used when printing a particular object (`missing`). Isn't there a more general theming system that we could use?
> 
> Anyway, regarding this particular issue, requiring users to set an environment variable seems suboptimal. Most of them won't even know they can do that. Is there a way to check whether `light_grey` gives the same color as the background, and switch to `dark_grey` in that case?

My proposal was to use medium gray by default and only if that collides with themes (which then are neither light nor dark but rather something in the middle, which I assue to be pretty rare) users are required to manualle chose a suitable color for de-highlighting. The problem is, that terminals just define highlighting-colors but no de-highlighting-colors.

**Comment by ronisbr** [comment]
> My proposal was to use medium gray by default and only if that collides with themes (which then are neither light nor dark but rather something in the middle, which I assue to be pretty rare) users are required to manually chose a suitable color for de-highlighting. The problem is, that terminals just define highlighting-colors but no de-highlighting-colors.

Notice that this is only possible if we assume a terminal capable of outputting 256 colors since the default 8 colors only contains bright black as gray. IMHO, I think the number of people using terminals that cannot output 256 colors should be minimal. Thus, it should be fine to select one of those colors:

<img width="972" alt="Captura de Tela 2023-11-13 às 09 33 28" src="https://github.com/JuliaData/DataFrames.jl/assets/1068295/e746993d-79bd-42db-b5e5-80d5de057e65">

Assuming 24-bit and selecting the RGB values is way more restrictive.

@fbruetting can you please test in your terminal what would be a good gray selection for those themes? To do so, open a terminal with your theme paste: `echo "\e[38;5;239mmissing"`, changing 239 to the color you are testing.

**Comment by fbruetting** [comment]
Of course, here it is for Solarized Dark and Light:

![Bildschirmfoto vom 2023-11-13 13-47-47](https://github.com/JuliaData/DataFrames.jl/assets/10520767/e4383779-86c4-45d2-9028-6be98c5b6bf6)![Bildschirmfoto vom 2023-11-13 13-48-18](https://github.com/JuliaData/DataFrames.jl/assets/10520767/7af2d6c5-0f92-45e3-b69f-97664dc2114a)

**Comment by ronisbr** [comment]
In the "default" themes, we will have:

<img width="695" alt="Captura de Tela 2023-11-13 às 10 12 41" src="https://github.com/JuliaData/DataFrames.jl/assets/1068295/15d85ab6-db64-4ec2-bb5e-b3a3ce2ba43e">

IMHO, the highlight will loose its purpose because it almost indistinguishable from the normal text (which usually is light gray instead of pure white).

**Comment by bkamins** [comment]
Side note: the same issue applies to the column type subheader.

**Comment by fbruetting** [comment]
Not having a good contrast is much less of a problem than not seeing those values at all, so I’d still prefer hardcoding medium gray.


Here’s a comparison of common themes for the colors 240-247 (all of these are shipped with Tilix):

### Solarized Light / Linux / Tango

![l02 - Solarized Light](https://github.com/JuliaData/DataFrames.jl/assets/10520767/dd79695b-46e2-40a5-b052-cce9816c19c4) ![l03 - Linux](https://github.com/JuliaData/DataFrames.jl/assets/10520767/7cbf24bd-8acf-46a1-aa7c-c722b022ee59) ![l07 - Tango](https://github.com/JuliaData/DataFrames.jl/assets/10520767/f0ed986f-52f7-446e-a289-7e90987b4d22)

### Solarized Dark / Material / Monokai Dark / Orchis / Yaru / Base16 Twilight dark

![d01 - Solarized Dark](https://github.com/JuliaData/DataFrames.jl/assets/10520767/8f3855d4-c76e-4e58-b794-d824f4bfd51e) ![d04 - Material](https://github.com/JuliaData/DataFrames.jl/assets/10520767/d82bd8e0-accb-429d-8897-d06cbd11c060) ![d05 - Monokai Dark](https://github.com/JuliaData/DataFrames.jl/assets/10520767/5125e2de-b94a-46ce-b686-c113cd0abbf7) ![d06 - Orchis](https://github.com/JuliaData/DataFrames.jl/assets/10520767/56194c01-eed6-43ef-912b-7848c15b19db) ![d08 - Yaru](https://github.com/JuliaData/DataFrames.jl/assets/10520767/a3ae0f6d-cc73-4841-9a4d-93bef62073d0) ![d09 - Base16 Twilight dark](https://github.com/JuliaData/DataFrames.jl/assets/10520767/6b618f2b-cae3-4c9b-86f8-e926668e6bf9)

**Comment by bkamins** [comment]
By the way - can you please confirm how the following renders under the color schemes we discuss? Thank you!
![image](https://github.com/JuliaData/DataFrames.jl/assets/6187170/a5e12e1d-920f-4739-b42c-93a30de2314c)
I am asking, because maybe we can use the gray that uses standard Julia in this printout.

**Comment by fbruetting** [comment]
That would lead to invisible text like in the Julia-issue referenced above.

Same order as above:

![Bildschirmfoto vom 2023-11-13 16-49-02](https://github.com/JuliaData/DataFrames.jl/assets/10520767/29344562-015c-41ef-be09-2bf5b9a74600)
![Bildschirmfoto vom 2023-11-13 16-49-09](https://github.com/JuliaData/DataFrames.jl/assets/10520767/52188fcd-1819-4a56-bdcc-19f575695d67)
![Bildschirmfoto vom 2023-11-13 16-49-14](https://github.com/JuliaData/DataFrames.jl/assets/10520767/c7d93baa-9ca6-49fa-b7da-f6e2f645ed2a)
![Bildschirmfoto vom 2023-11-13 16-49-21](https://github.com/JuliaData/DataFrames.jl/assets/10520767/32c44eb2-42c6-4964-92da-486764dba657)
![Bildschirmfoto vom 2023-11-13 16-49-26](https://github.com/JuliaData/DataFrames.jl/assets/10520767/77803160-bcf5-415a-8e09-597b5e927483)
![Bildschirmfoto vom 2023-11-13 16-49-33](https://github.com/JuliaData/DataFrames.jl/assets/10520767/25d74efd-6a87-485a-b7cd-5749e41c052d)
![Bildschirmfoto vom 2023-11-13 16-49-39](https://github.com/JuliaData/DataFrames.jl/assets/10520767/0cb2aed4-5ec6-4ed2-97d6-bf907b5d50a8)
![Bildschirmfoto vom 2023-11-13 16-49-49](https://github.com/JuliaData/DataFrames.jl/assets/10520767/ee2486a6-bf04-424c-b798-031cc9c55876)
![Bildschirmfoto vom 2023-11-13 16-49-54](https://github.com/JuliaData/DataFrames.jl/assets/10520767/e56a0deb-2e02-4efc-ae2a-0adc15246573)

**Comment by nalimilan** [comment]
What if we took a relatively dark grey like 238, which is not super readable in Solarized Dark but is at least discernible and better than the current situation on that theme? Anyway it's not hard to find out that these entries correspond to missing values, what matters is that you notice the cell isn't completely empty.

**Comment by fbruetting** [comment]
Relevant Solarized bug: https://github.com/altercation/solarized/issues/220

When I read this, there are MASSIVE complications and Solarized is to blame because of intentionally violating the specification. It might be better to let everything be as it is and let Solarized be broken, because when you’re fixing the theme for Solarized, all other themes suffer, because DF then hardcodes values in an environment where all other colors are variable.

Best solution would be to indroduce a config setting, which lets Solarized users alter that color if there’s need, then every other theme would work perfectly and there won’t be compromises at all – just a manual fix for an intentional spec violation. Seems the correct approach to me.

**Comment by ronisbr** [comment]
> Best solution would be to indroduce a config setting, which lets Solarized users alter that color if there’s need, then every other theme would work perfectly and there won’t be compromises at all – just a manual fix for an intentional spec violation. Seems the correct approach to me.

I totally agree. I had problems with solarized themes in the past for a similar reason. Thus, the option is to define a ENV variable that will change the colors in subheaders and in those highlighters.

**Comment by fbruetting** [comment]
And if that color is also hard-coded in Julia, then it would be even nicer if DF could refer to Julias grey color and Julia to implement a manual grey-color variable (if that’s not already possible). Then we’d have a Julia-global solution.

@nalimilan ? :smile:

**Comment by nalimilan** [comment]
I agree that Solarized seems broken. Given that it also affects Julia (and even more seriously than DataFrames), I'd prefer to find a more general solution there, e.g. which would allow overriding the colors that correspond to each named color. This should ideally be used by Crayons and we wouldn't have anything to do here.

**Comment by ronisbr** [comment]
> I agree that Solarized seems broken. Given that it also affects Julia (and even more seriously than DataFrames), I'd prefer to find a more general solution there, e.g. which would allow overriding the colors that correspond to each named color. This should ideally be used by Crayons and we wouldn't have anything to do here.

I think this is not the correct approach. Crayons.jl uses the escape sequences given the ANSI definitions. It should not modify the colors. The correct way to replace a color is to set the desired one in your terminal, by changing its theme. Of course, each application (DataFrames in this case) can provide a mechanism to tune the colors it uses if necessary. Since there is no definition of a "theme" in DataFrames because we only use few colors by default, I think an ENV to address this specific problem would not be bad.

**Comment by fbruetting** [comment]
It’s not always possible to modify the color palette. In GNOME Builder you for example can just select a theme but not modify it.

I think we should shrink `O(1 + n_pkgs)` to `O(1)` and have a global fix in Crayons then, I created an issue there:

**Comment by ronisbr** [comment]
> I think we should shrink O(1 + n_pkgs) to O(1) and have a global fix in Crayons then, I created an issue there:

No, it will not be global. There are packages, like Term.jl, (and Julia itself) that just output the escape sequences without using Crayons.jl at all. It will not have the intended behavior of a global fix. IMHO, if a theme is setting something that violates the ANSI specification (making a color equal to background when it just should not happen), the theme must be fixed.

Even inside PrettyTables.jl this will not be global. We have a text cell type called `AnsiTextCell` that allows to have decorated strings inside the tables cells. It does not use Crayons.jl, it parses the escape sequences as per ANSI specification. In this case, we will fix the `missing` in DataFrames.jl but those other cases will be completely inconsistent.

Allowing Crayons.jl (or anything) to manually change the color can lead to a lot of problems. Example: Package A uses `:red` to highlight an error. For some reason Package B overloads `:red` to be black. Package A will print invisible error messages. Hence, it would not be a simple task. It needs to be containerized so that each package has its own theme. This kind of thing requires changing Crayons.jl API or using global variables. In summary: it can be done, but it will not be simple, and I think it is really the wrong move due an error in a theme.

**Comment by ronisbr** [comment]
Btw, I was reading about the fixes in all the backlinks to that issue in solarized theme. Almost all involve changing the theme color in the application.

**Comment by fbruetting** [comment]
Hm, I understood nalimilan so as that the core color definition happens in Crayons and everything else refers to that.

You’re correct that the theme needs a fix, but reality is that this wasn’t fixed in 10 years and yet it’s usage went through the roof – so you can be sure that that won’t happen. There needs a fork to be established, which is a process of at least a decade, so we can’t do anything here, therefore the ugly ENV fix idea.

**Comment by ronisbr** [comment]
Yes, this is precisely my point! However, I also agree that solarized is indeed very popular. Hence, IMHO, the optimal approach (given the circumstances) would be an ENV in DataFrames that at the end is selecting the color for specific types of cells.

**Comment by ronisbr** [comment]
> We should strive for a Julia-global fix.

In this case, I do not agree because the only global fix will be fixing the theme, as many others are doing.

If we change the color definition in Crayons.jl, we fix this issue in DataFrames.jl, but the stack traces will continue to be unreadable.

IMHO, changing the color inside the application is like type-piracy. The terminal you are using is responsible to render the `light black` color. The `light black` color is defined by your theme. Hence, we can fix Julia stack traces manually (by changing the color), we can fix Crayons.jl, but any other package that prints raw escape sequences will be wrong.

**Comment by nalimilan** [comment]
Looking at https://github.com/JuliaLang/julia/issues/38730 I realize that Julia already has a workaround via e.g. `Base.text_colors[:light_black] = "\e[38;5;243m"`. We could use that value instead of `:dark_grey` (Julia doesn't have that color for some reason), given that by default `Base.text_colors[:light_black] == "\e[90m"`, i.e. it defaults to the value set by the theme.

I'd argue that the best place to do this would be even more upstream, i.e. in Crayons and similar packages, to ensure consistency. But anyway since this has to be set manually it won't affect users that don't want it so it won't make things worse than the current situation.
